components/sup/src/ctl_gateway/server.rs

```diff
@@ -481,14 +481,19 @@ pub fn run(listen_addr: SocketAddr, secret_key: String, mgr_sender: MgrSender) {
           let server =
               TcpListener::bind(&listen_addr).expect("Could not bind ctl gateway listen address!")
                                              .incoming()
-                                             .map(|tcp_stream| {
-                                                 let addr =
-                                                     tcp_stream.peer_addr().expect("Couldn't get \
-                                                                                    peer address!");
-                                                 let io = SrvCodec::new().framed(tcp_stream);
-                                                 let client = Client { handle: handle.clone(),
-                                                                       state:  state.clone(), };
-                                                 (client.serve(io), addr)
+                                             // Filter out instances where there is no peer address
+                                             // on the socket. peer_addr() can be Err when a client
+                                             // disregards the protocol such as from a `netcat -z`
+                                             // connection.
+                                             .filter_map(|tcp_stream| {
+                                                 if let Ok(addr) = tcp_stream.peer_addr() {
+                                                     let io = SrvCodec::new().framed(tcp_stream);
+                                                     let client = Client { handle: handle.clone(),
+                                                                           state:  state.clone(), };
+                                                     Some((client.serve(io), addr))
+                                                 } else {
+                                                     None
```

Comment:
It could be useful to log these instances. I doubt we're running into this a lot (otherwise we would have hit this issue a _waaaaay_ before now), but it might be nice to have some indication of when we _do_ run into it.

Reply:
@christophermaier great callout! I've added a debug in this commit https://github.com/habitat-sh/habitat/pull/7244/commits/179dc94c2dabb3eeb0f76e999761a9f1d357a2b1